### PR TITLE
Modify history contents/near endpoint to support new history scroller

### PR DESCRIPTION
**Pull request by jdavcs**
UPDATE: as per discussion here and on gitter, I've implemented this as one endpoint:
`/api/histories/{history-id}/contents/{direction}/{hid}/{limit}`

Direction can be `near|after|before`.

The limit is now the same regardless of direction. If `direction=near`, the up and down boundaries are calculated as follows:
`up = limit // 2 + 1 if limit //2 == 1 else limit // 2`
`down = limit // 2`
The "up" selection excludes the provided `hid`, the "down" selection includes it, so the total number is always `<= limit`. If one of the boundaries is greater than the available items in that direction, we do not shift the middle - i.e., if we have 10 items total (hids 1 through 10), `hid=9`, `limit=5`, then the result will be hids 7,8,9,10 - i.e., we will not move the selection window one hid down to accommodate the maximum limit of 5 (i.e., we will not select 6,7,8,9,10 instead). I think this would be the expected behavior in such a case. But this can be changed.

If `direction=after|before`, the provided `hid` is not included.

I've added API and unit tests. The unit tests verify assignment of total counts. The API tests check the results under several scenarios (even/odd limits; near/after/before directions with total counts being less than and greater than limits). I've verified this manually as well (although I don't know if it'll work for the scroller).

~~This is a new endpoint that's needed for the new history scroller.~~
~~No tests. It's a near-duplicate of the `contents/near` endpoint which has test coverage.~~ 
~~I excluded the `since` parameter because I suspect it's not needed here, but ***I'm not at all sure***.~~
I'm not adding a FastAPI controller (or more tests, for that matter) for the sake of getting this out ASAP so that this doesn't delay work on the new history.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:

Try to access `{your galaxy host}/api/histories/{history id}/contents/{direction}/{hid}/{limit}
If direction=near, hid=10 and limit=3, the data will contain hid=11, hid=12, hid=13.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/webapps/galaxy/api/history_contents.py
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/history_contents.py]:
```suggestion
    @expose_api_anonymous
```

File: lib/galaxy/webapps/galaxy/services/history_contents.py
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/services/history_contents.py]:
```suggestion
        return contents
```

**Review by davelopez — APPROVED** [review]
Looks good after aplying suggestions from @mvdbeek 

> I'm not adding a FastAPI controller (or more tests, for that matter) for the sake of getting this out ASAP so that this doesn't delay work on the new history.

No worries, I will follow up with those in https://github.com/galaxyproject/galaxy/pull/12578 after this is merged :)

**Comment by dannon** [comment]
I think it might make more sense to just add a directional param to 'near'?

**Comment by davelopez** [comment]
Yeah, we could add an Enum path parameter called `direction` to the service function that defaults to `near` and can be `before` and `after`.
Then call the service from the endpoint with the right direction /api/histories/{history_id}/contents/{direction}/{hid}/{limit} but I can do it in the next refactoring if this needs to be in soon to for the scroller support.

**Comment by dannon** [comment]
@davelopez I like that idea a lot as refactoring down the road.  For this right now, I think it makes sense to drop the extra endpoint added here and just and add the single direction param to the existing `contents_near` endpoint.  Then we still have since, up/down/etc.

**Comment by assuntad23** [comment]
I think this is going to return the opposite of what I need... I need hids that come before the given hid, and are below the given hid, because the history is in descending order. This reads like it'll be returning the higher hids that would sit above the current hid.

**Comment by dannon** [comment]
@assuntad23 Yep, we'll actually need both of those directional queries.

**Comment by ic4f** [comment]
What @assuntad23 meant is that she needs the "before" functionality - i.e., when we provide hid=10, the result should contain hid=9,8,7, not 11,12,13. I've made the change.

With regard to combining it in one endpoint, I must have misspoke at the meeting, because by "simpler and cleaner" I meant adding a separate endpoint, not adding more logic into the existing one - sorry!

I see the point of having a `direction` parameter, but I do have a few reservations. First, I am not sure we need separate `before` and `after` options: the `near` endpoint was supporting the virtual scroller (I don't know what else), but for the new infinite scroller we only need to expand in one direction - in decreasing hid order (@assuntad23 please correct me if I'm wrong). With that regard, the opposite direction is simply not needed at this point, I think (but I may be wrong).

My other reservation is that the `before/after` directions do not map cleanly onto `near`. If you have `near/5/2`, you get 4,5,6,7, but if you get `before/5/2` you'd expect 3,4 and for `after/5/2` -  6,7. So there's the `<` vs `<=` consideration. Also, in `near`, the results of the `after` query are [reversed](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/services/history_contents.py#L743) - which makes sense in the context of "near", but may or may not make sense in the context of a separate "after": who knows what a separate "after" endpoint may be called for. And then there's also the matter of the number of records: with `after` or `before` you'd get N records, but with `near, you'd get 2N: N-1 "before" + 1 + N "after". 

Also, the `near` endpoint has the `since` parameter handling. I may be wrong here, but I suspect it is not needed for the new scroller. (of course we can just not pass that argument.)

Overall, my suggestion is to have 2 separate endpoints: they do two separate, well-defined things, and they do them well. If we decide we need the opposite direction, we can switch the `before` to a `direction` parameter - but I would still not combine it with the logic in `near`. That said, if the consensus is to do one endpoint, I'm fine with that too.

**Comment by assuntad23** [comment]
@ic4f  Verifying that it is true that I only need the one direction (decreasing hid order) to move forward with the Scroller.

**Comment by dannon** [comment]
@ic4f We need before, after, and a locus (near) (or we can make two requests here, I guess).  It really seems to me like the best option is to have a single robust contents API with a parameter.

"they do two separate, well-defined things, and they do them well" -- they do the same thing, except one parameter is different?  That's what I'm seeing here.

**Comment by ic4f** [comment]
@dannon  OK, I'll implement it as one endpoint then.

**Comment by ic4f** [comment]
This is no longer a near-duplicate of an existing endpoint, so I've added API + unit tests. I'll add more details in the PR description.

File: lib/galaxy/schema/schema.py
Comment (davelopez) [inline lib/galaxy/schema/schema.py]: Since this is probably not going to be used by any other route than the `contents_near` I would put it in `lib/galaxy/webapps/galaxy/services/history_contents.py` but, as the `schema.py` module is growing a bit messy anyway is not a big deal I'm still trying to figure out what is the best place to put what :sweat_smile:
What do you think?

File: lib/galaxy/webapps/galaxy/api/history_contents.py
Comment (davelopez) [inline lib/galaxy/webapps/galaxy/api/history_contents.py:501]: Now that we are modifying this endpoint... I know Mason may have his reasons to abandon the q/qv syntax, I also found it strange at the beginning, but the thing is, the q/qv syntax can be modeled and documented in FastAPI, while for the new one there is no way to document every possible combination of parameters and also requires to do some [little hacks accessing the request query parameters directly](https://github.com/galaxyproject/galaxy/pull/12578/commits/e7d355bcfb80feecbe5700e85ecc0ee0805f8cb3#diff-d7789a8b5b8606bd040d59409203acc178c14fe4ec4d6b38e8955d87dc4ecea1R290-R301) while is changing how the rest of the endpoints behave.

But definitely, not something that should be addressed in this PR, I was commenting just to get a general opinion now that we have stumbled upon it again, and maybe I can address it in a follow-up PR if people agree.

**Review by davelopez — APPROVED** [review]
Excellent! especially all those awesome tests!

Just need to fix the relevant tests :)

File: lib/galaxy/schema/schema.py
Comment (jdavcs) [reply lib/galaxy/schema/schema.py]: I'll move it! I was actually going to ask you where you wanted me to put that in that monster file, but yes, moving it into the service file makes sense.

File: lib/galaxy_test/api/test_history_contents.py
Comment (assuntad23) [inline lib/galaxy_test/api/test_history_contents.py:753]: Great test cases!

File: lib/galaxy_test/api/test_history_contents.py
Comment (assuntad23) [inline lib/galaxy_test/api/test_history_contents.py:804]: The comments here are inaccurate, I think? The given hid was 5.

Comment (jdavcs) [inline lib/galaxy_test/api/test_history_contents.py]: You're right! The test is correct though. I'll adjust the comments - thanks!

File: lib/galaxy_test/api/test_history_contents.py
Comment (assuntad23) [inline lib/galaxy_test/api/test_history_contents.py]:
```suggestion
            assert result[0]['hid'] == 7  # hid + 2
            assert result[1]['hid'] == 6  # hid + 1
            assert result[2]['hid'] == 5  # hid
            assert result[3]['hid'] == 4  # hid - 1
```

**Review by assuntad23 — CHANGES_REQUESTED** [review]

File: lib/galaxy/webapps/galaxy/api/history_contents.py
Comment (jdavcs) [reply lib/galaxy/webapps/galaxy/api/history_contents.py:501]: Marking this as resolved, as the discussion moved to matrix.

**Comment by ic4f** [comment]
I've reverted the last commit that added error handling for incorrect direction parameter. This route precedes several other routes of the same form: `/api/histories/{history-id}/contents/{some-parameter}/...`, so the error handling, obviously, breaks all subsequent endpoints (as listed in buildapp.py). I don't think moving this endpoint to the end of the list solves the bigger issue (what about error handling of the params in the other endpoints), so I'm leaving it as is.

**Comment by ic4f** [comment]
I've added 3 routes to buildapp.py, hardcoding the direction into the route (after/before/near). I expect this will fix the failing tests. 

The `direction` parameter was causing the wrong endpoint to be be selected, which had the same (or same up to some point) structure, so the wrong controller was called, the arguments where mixed up and `direction`, I think, was interpreted as `history_content_id` (unless it was yet another endpoint, but the issue is the same). I'm not sure how else to fix this: moving this route up would, clearly, break the subsequent routes that have the same structure (as far as I know there is no "fall through" in route selection). There might be some magic in route precedence rules, but I'm not familiar with it. 

Any help or advice is appreciated, of course. But if this solution is good enough, let's merge it, because it may be holding up work on the scroller.

**Comment by ic4f** [comment]
So, the `direction` parameter won't work, at least in the present setup. The new route is conflicting with some other route that has the same structure (now that "near" has been turned into a parameter); I am not sure which one exactly (I can't tell from all the path prefixes in buildapp.py), but what's certain is that this route is being selected incorrectly, the controller is called with the wrong arguments, and `direction` is interpreted as `history_content_id` ("dataset_collection", I think). Which is why tests fail. (and rightly so!) 

Moving the endpoint won't work, unless there's some magic in route precedence rules that I'm unfamiliar with. The solutions I see are (a) change the endpoint to make it unique; and (b) follow suggestion from @davelopez and have multiple endpoints with one service method handling all of them. Option b seems like the lesser of two hacks and, definitely less painful in my estimate. I haven't given any thought to what a new unique endpoint could be, and I don't think that should be taken lightly, at least not without considering what that would add to the set of all endpoints. So, out of these two, my suggestion is option b. Maybe there are better options.

Any opinion, help or advice is appreciated, of course. But if this solution is good enough, and if the tests pass already :crossed_fingers: , let's merge it, because it may be holding up work on the scroller.

Ping @dannon, @assuntad23

**Comment by davelopez** [comment]
I'm not really familiar with this `routes` library, but looking at the documentation I found that you can restrict a path parameter to certain values using *requirements* ([link to documentation](https://routes.readthedocs.io/en/latest/setting_up.html#requirements)). That is what is actually used for example here:

https://github.com/galaxyproject/galaxy/blob/f23e8e1681d29db046c4a44f12e3ba580e6b73a1/lib/galaxy/webapps/galaxy/buildapp.py#L605

So maybe something similar to this might work:
```
webapp.mapper.connect("/api/histories/{history_id}/contents/{direction:near|before|after}/{hid}/{limit}",
                          action="contents_near",
                          controller='history_contents',
                          conditions=dict(method=["GET"]))
```

Anyway, since all tests are passing, I will go ahead and merge. These legacy endpoints are going to be replaced soon™ so I guess it's ok like this.

Thanks again for the nice test coverage!

**Comment by ic4f** [comment]
@davelopez thank you for looking this up and for the link to the docs! And thanks for the merge - given that these are going to be replaced soon, I think this should be sufficient.

**Pull request merged**
